src/tests/unittests/asn1.c

```diff
@@ -0,0 +1,275 @@
+		int rv; \
+	\
+		rv = sc_asn1_decode_integer(data, datalen, &value, 1); \
+		assert_int_equal(rv, error); \
+		/* but we can parse them without the strict checking */ \
+		rv = sc_asn1_decode_integer(data, datalen, &value, 0); \
+		assert_int_equal(rv, SC_SUCCESS); \
+		assert_int_equal(value, int_value); \
+	}
+
+/* Data are without the Tag (0x02) and Length */
+/* Positive test cases, mostly corner cases */
+TORTURE_INTEGER(zero, "\x00", 0)
+TORTURE_INTEGER(one, "\x01", 1)
+TORTURE_INTEGER(minus_one, "\xFF", -1)
+TORTURE_INTEGER(padded_128, "\x00\x80", 128)
+TORTURE_INTEGER(max2, "\x7F\xFF", 32767)
+TORTURE_INTEGER(min2, "\x80\x00", -32768)
+
+#if INT_MAX == 2147483647
+TORTURE_INTEGER(max4, "\x7F\xFF\xFF\xFF", 2147483647)
+TORTURE_INTEGER(min4, "\x80\x00\x00\x00", -2147483648)
+#else
+TORTURE_INTEGER_ERROR(max4, "\x7F\xFF\xFF\xFF", SC_ERROR_NOT_SUPPORTED)
+TORTURE_INTEGER_ERROR(min4, "\x80\x00\x00\x00", SC_ERROR_NOT_SUPPORTED)
+#endif
+
+/* Negative test cases */
+TORTURE_INTEGER_ERROR(null, "", SC_ERROR_INVALID_ASN1_OBJECT)
+TORTURE_INTEGER_ERROR(over, "\x7F\xFF\xFF\xFF\xFF", SC_ERROR_NOT_SUPPORTED)
+
+/* Tests fail in strict mode, but work otherwise */
+TORTURE_INTEGER_NONSTRICT(padded_zero, "\x00\x00", SC_ERROR_INVALID_ASN1_OBJECT, 0)
+TORTURE_INTEGER_NONSTRICT(padded_one, "\x00\x01", SC_ERROR_INVALID_ASN1_OBJECT, 1)
+TORTURE_INTEGER_NONSTRICT(padded_minus_one, "\xFF\xFF", SC_ERROR_INVALID_ASN1_OBJECT, -1)
+TORTURE_INTEGER_NONSTRICT(padded_127, "\x00\x7F", SC_ERROR_INVALID_ASN1_OBJECT, 127)
+
+/*
+ * Test undefined behavior of negative INTEGERS handling.
+ * https://oss-fuzz.com/testcase-detail/5125815506829312
+ *
+ * The issue was not actually the size of the integers, but that first
+ * negative value wrote ones to the whole integer and it was not possible
+ * to shift values afterward.
+ */
+TORTURE_INTEGER(negative, "\xff\x20", -224)
+
+#define TORTURE_BIT_FIELD(name, asn1_data, int_value) \
+	static void torture_asn1_bit_field_## name (void **state) \
+	{ \
+		u8 data[] = asn1_data; \
+		size_t datalen = sizeof(data) - 1; \
+		unsigned int value = 0; \
+		size_t value_len = sizeof(value); \
+		int rv; \
+	\
+		rv = decode_bit_field(data, datalen, &value, value_len); \
+		assert_int_equal(rv, SC_SUCCESS); \
+		assert_int_equal(value, int_value); \
+	}
+#define TORTURE_BIT_FIELD_ERROR(name, asn1_data, error) \
+	static void torture_asn1_bit_field_## name (void **state) \
+	{ \
+		u8 data[] = asn1_data; \
+		size_t datalen = sizeof(data) - 1; \
+		unsigned int value = 0; \
+		size_t value_len = sizeof(value); \
+		int rv; \
+	\
+		rv = decode_bit_field(data, datalen, &value, value_len); \
+		assert_int_equal(rv, error); \
+	}
+/* Without the Tag (0x03) and Length */
+/* Simple value 1 */
+TORTURE_BIT_FIELD(one, "\x07\x80", 1)
+/* This is the last value that can be represented in the unsigned int */
+TORTURE_BIT_FIELD(uint_max, "\x00\xff\xff\xff\xff", UINT_MAX)
+/* Valid padding */
+TORTURE_BIT_FIELD(padding, "\x01\xfe", 127)
```

Comment:
wait, so the decoded bitstring doesn't remember the size of the input string?
so `\x00\x00\x00` and `\x00\x00` are equal?!

Reply:
Yes. In this case, when we request the decoding to integer type, there is no way how we can remember also the length. This is described in the header of the function in 

https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/asn1.c#L663

There is also API to decode the BIT STRINGs into byte fields:

https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/asn1.c#L569

Reply:
ah, ok then, thanks for the references